resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.resolver.dns;
+
+import io.netty.channel.EventLoop;
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.UnstableApi;
+
+import java.net.InetSocketAddress;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+
+import static io.netty.util.internal.ObjectUtil.*;
+
+/**
+ * Default implementation of {@link AuthoritativeDnsServerCache}, backed by a {@link ConcurrentMap}.
+ */
+@UnstableApi
+public class DefaultAuthoritativeDnsServerCache implements AuthoritativeDnsServerCache {
+
+    private final int minTtl;
+    private final int maxTtl;
+    private final Comparator<InetSocketAddress> comparator;
+    private final Cache<InetSocketAddress> resolveCache = new Cache<InetSocketAddress>() {
+        @Override
+        protected boolean shouldReplaceAll(InetSocketAddress entry) {
+            return false;
+        }
+
+        @Override
+        protected boolean equals(InetSocketAddress entry, InetSocketAddress otherEntry) {
+            return entry.getHostName().equalsIgnoreCase(otherEntry.getHostName());
```

Comment:
This should never happen as we will only try to add if the hostname is resolved:

https://github.com/netty/netty/pull/8177/files#diff-7ab1cd3db010ccc37c01987e522675caR104

That said let me use getHostString when on java7+

Reply:
IIUC we add unresolved entries and resolved entries to the `AuthoritativeDnsServerCache`. The reverse lookup occurs if the address is created via an IP, so seems like this would actually be an issue for the resolved cases.

Reply:
we should only add entries which have a hostname

Reply:
got it ... we always have/use the name server name (and not just the IP) 👍